### PR TITLE
WFLY-5550 Distributable Web subsystem

### PR DESCRIPTION
**Pull request by pferraro**
JIRA:
https://issues.jboss.org/browse/WFLY-5550
Design doc:
https://github.com/wildfly/wildfly-proposals/pull/111

Since this PR adds a new extension to the configuration, it requires new wildfly-core release that contains https://issues.jboss.org/browse/WFCORE-4137.

### CODE REVIEW

File: clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd
Comment (claudio4j) [inline clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd]: Typo: mangement

File: clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd
Comment (claudio4j) [inline clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd]: Typo: mangement

**Comment by pferraro** [comment]
@jmesnil Please remove the core-upgrade-need flag from this PR, as it was reworked not to require WFCORE-3939.

File: web-common/src/main/java/org/jboss/as/web/session/SharedSessionManagerConfig.java
Comment (jamezp) [inline web-common/src/main/java/org/jboss/as/web/session/SharedSessionManagerConfig.java]: Why the change to an object?

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java
Comment (jamezp) [inline clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java]: This could just use `DistributableWebDeploymentSchema.values()`.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentParsingProcessor.java
Comment (jamezp) [inline clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentParsingProcessor.java]: This could just use `DistributableWebDeploymentSchema.values()`

File: undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemAdd.java
Comment (jamezp) [inline undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemAdd.java]: This could just use `SharedSessionConfigSchema.values()`.

**Review by jamezp — CHANGES_REQUESTED** [review]
One thing I didn't see here were any transformers which seems like they should exist since new attributes have been added.

What happens to old `jboss-web.xml` configurations? Do they still work or are changes required?

We should open an email to the layered products list about this as well since some things that are semi-public are being removed.

**Comment by pferraro** [comment]
@jamezp 
> One thing I didn't see here were any transformers which seems like they should exist since new attributes have been added.

This change is not adding any new attributes, but rather is adding a whole subsystem - so I'm not sure what I'd be transforming.  How are new subsystems generally handled with respect to compatibility?

> What happens to old jboss-web.xml configurations?  Do they still work or are changes required?

Old jboss-web.xml configuration (i.e. those specifying &lt;replication-config/&gt;) will continue to work as they always have (as validated by existing integration tests).
When processing a distributed web application deployment, if the JBossWebMetaData contains a non-null ReplicationConfig, we generate a legacy SessionManagementProvider from it, which takes precedence over any deployment-specific or default SessionManagementProvider:
https://github.com/pferraro/wildfly/blob/web/clustering/web/undertow/src/main/java/org/wildfly/clustering/web/undertow/session/UndertowSessionManagementProviderFactory.java#L56

> We should open an email to the layered products list about this as well since some things that are semi-public are being removed.

This change isn't removing anything.  To which changes are you referring?

File: clustering/common/src/main/java/org/jboss/as/clustering/controller/persistence/AttributeXMLBuilderOperator.java
Comment (pferraro) [reply clustering/common/src/main/java/org/jboss/as/clustering/controller/persistence/AttributeXMLBuilderOperator.java]: Yes.  We need to ensure that the attributes that are parsed/marshalled as elements are added to the PersistentResourceXMLBuilder last.  Otherwise, the generic parser/write falls on its face.

File: clustering/web/extension/pom.xml
Comment (pferraro) [reply clustering/web/extension/pom.xml]: Nope.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebExtension.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebExtension.java]: EnumSet.allOf(...) iteration is more efficient than Enum.values().

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceServiceHandler.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceServiceHandler.java]: EnumSet.allOf(...) iteration is more efficient than Enum.values().

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/undertow/main/module.xml
Comment (pferraro) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/undertow/main/module.xml]: org.wildfly.clustering.web.container is a new module being introduced by this PR.
org.wildfly.clustering.web.api is actually an obsolete module dependency.

Comment (jamezp) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/undertow/main/module.xml]: Can it not be optional? This now makes it a requirement which means it can't be stripped down.

File: web-common/src/main/java/org/jboss/as/web/session/SharedSessionManagerConfig.java
Comment (pferraro) [reply web-common/src/main/java/org/jboss/as/web/session/SharedSessionManagerConfig.java]: 2 reasons:
1. An unlimited number of active sessions is more clearly represented by a null value.
2. To match the semantics of the corresponding JBossWebMetaData property:
https://github.com/jboss/metadata/blob/master/web/src/main/java/org/jboss/metadata/web/jboss/JBossWebMetaData.java#L157

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java]: EnumSet.allOf(...) iteration is more efficient than Enum.values().

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentParsingProcessor.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentParsingProcessor.java]: EnumSet.allOf(...) iteration is more efficient than Enum.values().

File: undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemAdd.java
Comment (pferraro) [reply undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemAdd.java]: EnumSet.allOf(...) iteration is more efficient than Enum.values().

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceServiceHandler.java
Comment (jamezp) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceServiceHandler.java]: Interesting. I guess I was thinking more of the memory overhead for cloud, but I guess more than likely it would be wrapped in some kind of ArrayList anyway.

**Comment by jamezp** [comment]
Ah right yes a new subsystem that explains it. I thought for some reason we rejected the add, but I guess it would fail first with the extension being added so maybe there isn't anything we can do. @jmesnil do we do anything special with rejecting ops to add new subsystems?

The concern I had with the "public" part being dropped was mainly https://github.com/wildfly/wildfly/pull/11476/files#diff-3e567146c3f59369e394436ca993f3f3L56. I would hope other projects, like Keycloak, aren't using that however I don't know that we can be sure they aren't.

**Comment by pferraro** [comment]
> The concern I had with the "public" part being dropped was mainly https://github.com/wildfly/wildfly/pull/11476/files#diff-3e567146c3f59369e394436ca993f3f3L56. I would hope other projects, like Keycloak, aren't using that however I don't know that we can be sure they aren't.

Ah, I see.  Are deployment attachments genuinely public?  I removed this assuming it was only used internally by the undertow subsystem (to provide support for a non-spec feature).  I can certainly keep the attachment around (maybe deprecate its key?) in case it is referenced elsewhere.

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/undertow/main/module.xml
Comment (pferraro) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/undertow/main/module.xml]: No, it cannot.  The org.wildfly.clustering.web.container module only contains a set of interfaces - which were refactored from the undertow subsystem itself.  This module is what allows the org.wildfly.clustering.web.extension module (containing the distributable web subsystem which configures the distributed session manager implementations) to avoid a direct dependency on the undertow subsystem, while allowing the undertow subsystem to avoid a direct dependency on any clustering implementation module.

File: undertow/src/main/java/org/wildfly/extension/undertow/Server.java
Comment (pferraro) [reply undertow/src/main/java/org/wildfly/extension/undertow/Server.java]: Fixed.

**Comment by bstansberry** [comment]
+1 to deprecating the attachment key; thanks. It's hard to tell who's going to be using stuff, so if a subsystem wasn't clearly meant to be a leaf in any dependency graph it's good to be cautious.

**Comment by bstansberry** [comment]
I looked at this over the weekend and

1) All the commits except "WFLY-5550 Distributable Web subsystem" look fine.
2) I didn't see anything wrong in the "WFLY-5550 Distributable Web subsystem" commit. It's huge though so I can't say I reviewed it closely. But probably as well as any general code review can do.  I think the key for that is testing, not code review.

I suggest that if anyone else besides James and I looks at this that they focus on that one commit.

File: feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml
Comment (yersan) [inline feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml]: The following modules are not used to cover a direct dependency of this module:
- org.wildfly.security.elytron-private
- org.wildfly.clustering.ee.infinispan
- javax.servlet.api
- org.jboss.threads
- org.wildfly.common
- org.wildfly.clustering.ee.spi
- org.wildfly.clustering.web.undertow
- org.jboss.modules
- org.infinispan.commons
- org.wildfly.clustering.web.container

I checked only direct dependencies, I have not checked if they are used to supply a service loaded by this module or if they are required due to an external activation. That requires a wide knowledge of this subsystem, but maybe the list of required dependencies can be reduced using the above list.

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/web/container/main/module.xml
Comment (yersan) [inline servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/web/container/main/module.xml]: The following modules are not used to cover a direct dependency of this module:
- org.jboss.as.controller
- org.wildfly.security.elytron-private
- org.jboss.dmr
- org.wildfly.common
- org.jboss.as.server
- org.jboss.logging
- org.wildfly.clustering.service
- org.wildfly.clustering.web.undertow

I checked only direct dependencies, I have not checked if they are used to supply a service loaded by this module or if they are required due to an external activation. That requires a wide knowledge of this subsystem, but maybe the list of required dependencies can be reduced using the above list.

File: feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml
Comment (pferraro) [reply feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml]: Thanks for the analysis - some of these are certainly required, but I'll verify the others.

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/web/container/main/module.xml
Comment (pferraro) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/web/container/main/module.xml]: Thanks for the analysis - some of these are certainly required, but I'll verify the others.

**Review by jamezp — COMMENTED** [review]
Overall this seems okay to me. I do think the comment about the `parentUnit.getParent().getName()` is likely just a mistake. The two comments on the potential race conditions is likely not an issue, but I thought I'd point it out just in case it is.

This is likely not a huge deal either, but I see some stuff that was in Undertow has been moved to a shared module. We _may_ need to check with the layered products to make sure they aren't using them or at least just inform them they've moved. I feel it's unlikely they are used outside WildFly though.

Consider this approved by me assuming the `parentUnit.getParent().getName() is correct or gets fixed.

**Comment by jamezp** [comment]
Oh I didn't notice any tests for this either besides subsystem tests. I also didn't see any transformers to reject the new resources.

**Comment by jamezp** [comment]
retest this please

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java]: Oh man - good catch.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/InfinispanSessionManagementServiceConfigurator.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/InfinispanSessionManagementServiceConfigurator.java]: These fields are volatile since they are read by a different thread than the thread that sets them.  The getters in question are invoked after the configure(...) method is called (i.e. when then service that exposes the interface containing those getters starts), so there is no potential race condition.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SessionManagementServiceConfigurator.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SessionManagementServiceConfigurator.java]: Same as above.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java]: Fixed.

**Comment by pferraro** [comment]
@jamezp 

> Oh I didn't notice any tests for this either besides subsystem tests.

https://github.com/wildfly/wildfly/pull/11476/commits/772a97a5ec1e30c987d79754021a0eee0ca9aca6

> I also didn't see any transformers to reject the new resources.

Since this is a new subsystem there is no previous version of the subsystem for which to register such transformers.
What is to be done in this case?

**Comment by bstansberry** [comment]
> Since this is a new subsystem there is no previous version of the subsystem for which to register such transformers. What is to be done in this case?

There's no transforming possible except in the rare case where everything in the new subsystem maps to something in an old subsystem.

New extensions should be added to the host-exclude configs:

feature-pack/src/main/resources/configuration/domain/template.xml
galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml
servlet-feature-pack/src/main/resources/configuration/domain/template.xml
servlet-galleon-pack/src/main/resources/feature_groups/domain.xml

File: clustering/web/extension/src/main/resources/subsystem-templates/distributable-web.xml
Comment (yersan) [inline clustering/web/extension/src/main/resources/subsystem-templates/distributable-web.xml]: This is pretty minor, the order of the xml attributes generated by galleon is not defined, but it seems it's always the same, would you mind using this order here?
`<infinispan-session-management name="default" granularity="SESSION" routing="OWNER" cache-container="web"/>`
That matches with the one generated by galleon. It will reduce differences based on the position of the attributes when we are comparing legacy-builds and builds done via galleon.

File: galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml
Comment (yersan) [inline galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml]: I suggest keeping the alphabetical order here and put org.wildfly.extension.clustering.web exclusion as the first in the list

**Review by yersan — COMMENTED** [review]
I did a comparison between legacy-builds and standard builds and, besides previous minor comments, the org.wildfly.extension.clustering.web is missing in the following configuration examples used in legacy builds:

feature-pack/src/main/resources/configuration/examples/subsystems-activemq-colocated.xml
feature-pack/src/main/resources/configuration/examples/subsystems-genericjms.xml
feature-pack/src/main/resources/configuration/examples/subsystems-picketlink.xml
feature-pack/src/main/resources/configuration/examples/subsystems-jts.xml
feature-pack/src/main/resources/configuration/examples/subsystems-rts.xml
feature-pack/src/main/resources/configuration/examples/subsystems-xts.xml

**Comment by pferraro** [comment]
> There's no transforming possible except in the rare case where everything in the new subsystem maps to something in an old subsystem.

Thanks for clarifying.  That's what I figured.
 
> New extensions should be added to the host-exclude configs:
> 
> feature-pack/src/main/resources/configuration/domain/template.xml
> galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml

This was already done in this commit: https://github.com/wildfly/wildfly/pull/11476/commits/4aa9b0b7e6a3dd1603d0c4f4648514e762018cf9

File: clustering/web/extension/src/main/resources/subsystem-templates/distributable-web.xml
Comment (pferraro) [reply clustering/web/extension/src/main/resources/subsystem-templates/distributable-web.xml]: :-1:  Given that the attribute ordering in galleon is undefined - any change to the set of attributes will likely completely change the ordering within xml.  I don't think it is realistic to ask each subsystem maintainer to keep these in sync.

File: galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml
Comment (pferraro) [reply galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml]: Oh - I didn't realize these were lexicographically ordered.  I figured that adding this subsystem to the end would make reviewing diffs easier.  I'll fix.

Comment (yersan) [reply galleon-pack/src/main/resources/feature_groups/domain-host-excludes.xml]: We are trying to follow an alphabetical order here, but really it's not a requirement, what is useful is to follow the same order than the legacy template: feature-pack/src/main/resources/configuration/domain/template.xml. It is just it helps with comparison between standard and legacy builds.

**Review by jamezp — APPROVED** [review]

File: docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc
Comment (tommaso-borgato) [inline docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc]: @pferraro what about adding some cli scripts? :

The "foo" session management profile can be created on the server like this:
----
/subsystem=distributable-web/infinispan-session-management=sm_offload:add(cache-container=web,cache=foo,granularity=SESSION)
----
Note that a cache named "foo" in cache container "web" must exist in subsystem "infinispan"; if it did not exist already, it can be created like this:
----
/subsystem=infinispan/cache-container=web/distributed-cache=foo:add()
----

for "affinity=none":
----
/subsystem=distributable-web/infinispan-session-management=sm_offload/affinity=none:add()
----

for "affinity=local":
----
/subsystem=distributable-web/infinispan-session-management=sm_offload/affinity=local:add()
----

for "affinity=primary-owner":
----
/subsystem=distributable-web/infinispan-session-management=sm_offload/affinity=primary-owner:add()
----

File: docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc
Comment (tommaso-borgato) [inline docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc]: @pferraro what about adding the other tags? (I had to execute cli to find out):

```

for "affinity=none" replace tag `<local-affinity/>` with tag:
[source,xml]
----
<no-affinity/>
----

for "affinity=primary-owner" replace tag `<local-affinity/>` with tag:
[source,xml]
----
<primary-owner-affinity/>
----

```

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/async/AsyncServletTestCase.java
Comment (tommaso-borgato) [inline testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/async/AsyncServletTestCase.java]: This test uses the same jboss-all_fine.xml file as FineWebFailoverTestCase.java, what about using a new one for covering the reference to an existing server profile from a jboss-all.xml file? (at present we have it from distributable-web.xml file in testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/persistence/distributable-web.xml)

Comment (pferraro) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/async/AsyncServletTestCase.java]: OK - I can do that.  That said, I'm not overly concerned with which configuration file contains the config, as the namespace has only 1 parser, but this is trivial enough to cover.

Comment (pferraro) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/async/AsyncServletTestCase.java]: Done.

**Pull request merged**
